demos/complementary-color-prediction-eager/complementary-color-prediction-eager.ts

```diff
@@ -0,0 +1,415 @@
+  }
+
+  /**
+   * Constructs weights and generates data. Call this method before training.
+   */
+  setupSession(): void {
+    // Generate and initialize the weight variables for fully connected layers
+    this.setupModel();
+    // Generate the data that will be used to train the model.
+    this.generateTrainingData(this.sampleSize);
+  }
+
+  /**
+   * Constructs the weight variables of the model. Call this method before
+   * training with fully connected layers.
+   */
+  setupModel(): void {
+    // Create 3 fully connected layers, each with half the number of nodes of
+    // the previous layer. The first one has 64 nodes.
+    this.createFullyConnectedLayerWeights(3, 64, 0);
+
+    // Create fully connected layer 1, which has 32 nodes.
+    this.createFullyConnectedLayerWeights(64, 32, 1);
+
+    // Create fully connected layer 2, which has 16 nodes.
+    this.createFullyConnectedLayerWeights(32, 16, 2);
+
+    // Create fully connected output layer, which has 3 nodes.
+    this.createFullyConnectedLayerWeights(16, 3, 3);
+  }
+
+  /**
+   * Trains one batch for one iteration. Call this method multiple times to
+   * progressively train.
+   *
+   * If shouldFetchCost is true, returns the mean cost across examples in the
+   * batch. Otherwise, returns null.
+   */
+  train1Batch(data: DataInterface, shouldFetchCost: boolean): dl.Scalar {
+    // Train 1 batch.
+    return dl.tidy(() => {
+      const { input, target } = data;
+
+      return this.optimizer.minimize(() => {
+        const prediction = this.predict(input);
+        return this.loss(prediction, dl.tensor2d(
+          target, [target.length, target[0].length]));
+      }, shouldFetchCost);
+    });
+  }
+
+  /**
+   * Trains all batches within the training epoch for one iteration. Call
+   * this method multiple times to progressively train. Calling this function
+   * with shouldFetchCost equal to true transfers data from the GPU in order
+   * to obtain the current loss on training data.
+   *
+   * If shouldFetchCost is true, returns the mean cost across examples in the
+   * all the batches within the training epoch. Otherwise, returns null. We
+   * should only retrieve the cost now and the because doing so requires
+   * transferring data from the GPU.
+   */
+  train1Step(step: number, shouldFetchCost: boolean): number {
+    // Every 42 steps, lower the learning rate by 15%.
+    const learningRate =
+      this.learningRate * Math.pow(0.85, Math.floor(step / 42));
+    this.optimizer.setLearningRate(learningRate);
+
+    // Train 1 batch.
+    return dl.tidy(() => {
+      let cost: dl.Scalar = dl.scalar(0);
+      const { input, target } = this.data;
+      const lastBatchIndex = input.length - this.batchSize;
+      for (let i = 0; i <= lastBatchIndex; i += this.batchSize) {
+        const batchData = {
+          input: input.slice(i, i + this.batchSize),
+          target: target.slice(i, i + this.batchSize),
+        };
+        if (shouldFetchCost) {
```

Comment:
I think, it should be the following:
```
      const batchCost = this.train1Batch(batchData, shouldFetchCost);
      if(shouldFetchCost){
        cost = cost.add(batchCost);
      }
```
Otherwise train1Batch will only be called when shouldFetchCost is set to true.